### ISSUE TITLE
[Feature]: Evaluate Profile-Guided Optimization (PGO) and LLVM BOLT

### ISSUE THREAD

**Issue by zamazan4ik** [oceanbase/oceanbase#1580]
Hi!

Recently I did many Profile-Guided Optimization (PGO) benchmarks on multiple projects (including many databases like PostrgreSQL, MongoDB, Redis, and many others) - the results are available [here](https://github.com/zamazan4ik/awesome-pgo), all databases results are located [here](https://github.com/zamazan4ik/awesome-pgo#databases). So that's why I think it's worth trying to apply PGO to OceanBase.

I can suggest the following things to do:
* Evaluate PGO's applicability and benchmark results to OceanBase.
* If PGO helps to achieve better performance - add a note to OceanBase's documentation about that. In this case, users and maintainers will be aware of another optimization opportunity for OceanBase.
* Provide PGO integration into the build scripts. It can help users and maintainers easily apply PGO for their own workloads.
* PGO-optimize prebuilt OceanBase binaries (if it's possible to prepare a generic-enough test workload).

Here are some PGO integration examples in the existing build scripts in other projects:
* Rustc: a CI [script](https://github.com/rust-lang/rust/blob/master/src/ci/stage-build.py) for the multi-stage build
* GCC:
  - Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback" (even AutoFDO build is supported)
  - A [part](https://github.com/gcc-mirror/gcc/blob/4832767db7897be6fb5cbc44f079482c90cb95a6/configure#L7818) in a "wonderful" `configure` script 
* Clang: [Docs](https://llvm.org/docs/HowToBuildWithPGO.html) 
* Python: 
  - CPython: [README](https://github.com/python/cpython#profile-guided-optimization)
  - Pyston: [README](https://github.com/pyston/pyston#building)
* Go: [Bash script](https://github.com/golang/go/blob/master/src/cmd/compile/profile.sh)
* V8: [Bazel flag](https://github.com/v8/v8/blob/main/BUILD.gn#L184)
* ChakraCore: [Scripts](https://github.com/chakra-core/ChakraCore/tree/master/Build/scripts/pgo)
* Chromium: [Script](https://chromium.googlesource.com/chromium/src/build/config/+/refs/heads/main/compiler/pgo/BUILD.gn)
* Firefox: [Docs](https://firefox-source-docs.mozilla.org/build/buildsystem/pgo.html)
   - Thunderbird has PGO support too
* PHP - [Makefile command](https://github.com/php/php-src/blob/master/build/Makefile.global#L138) and old Centminmod [scripts](https://github.com/centminmod/php_pgo_training_scripts)
* MySQL: [CMake script](https://github.com/mysql/mysql-server/blob/8.0/cmake/fprofile.cmake)
* YugabyteDB: [GitHub commit](https://github.com/yugabyte/yugabyte-db/commit/34cb791ed9d3d5f8ae9a9b9e9181a46485e1981d)
* FoundationDB: [Script](https://github.com/apple/foundationdb/blob/1a6114a66f3de508c0cf0a45f72f3687ba05750c/contrib/generate_profile.sh)
* Zstd: [Makefile](https://github.com/facebook/zstd/blob/dev/programs/Makefile#L232)
* [Foot](https://codeberg.org/dnkl/foot): [Scripts](https://codeberg.org/dnkl/foot/src/branch/master/pgo)
* Windows Terminal: [GitHub PR](https://github.com/microsoft/terminal/pull/10071)
* Pydantic-core: [GitHub PR](https://github.com/pydantic/pydantic-core/pull/741)

Some PGO documentation examples:
* ClickHouse: https://clickhouse.com/docs/en/operations/optimizing-performance/profile-guided-optimization
* Databend: https://databend.rs/doc/contributing/pgo
* Vector: https://vector.dev/docs/administration/tuning/pgo/
* Nebula: https://docs.nebula-graph.io/3.5.0/8.service-tuning/enable_autofdo_for_nebulagraph/
* GCC: Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback" (even AutoFDO build is supported)
* Clang:
  - https://llvm.org/docs/HowToBuildWithPGO.html
  - https://llvm.org/docs/AdvancedBuilds.html

After PGO, I can suggest evaluating [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md) as an additional optimization step after PGO.

**Comment by caifeizhi** [comment]
Thank you for your suggestion. 
Currently, the code in OceanBase already supports FGO, and you can find the code at: https://github.com/oceanbase/oceanbase/blob/e4e09ecfb65d55d6225b080f6f2069159c94c1c7/cmake/Env.cmake#L55 You are welcome to provide any further optimization suggestions for this. 
Additionally, we would appreciate it if you could add our repository to the awesome repository related to FGO.

**Comment by zamazan4ik** [comment]
> Currently, the code in OceanBase already supports FGO, and you can find the code at

Thanks for pointing to the script! I didn't find it because I had been searching for PGO instead of FDO... :)

I have some questions about the current FDO implementation in OceanBase:
1. Do you have the benchmark results of enabling FDO on OceanBase? I am interested in comparing Release vs Release + PGO OceanBase versions. So I will be able to put your benchmarks to my repo as an example.
2. I see you use Sampling (AutoFDO) approach. Did you try to use the Instrumented FDO instead? If yes, could you please describe why you chose AutoFDO instead? My guess is that Instrumented FDO has just too big performance overhead compared to AutoFDO but it's just my guess.
3. How did you collect the committed to the repo FDO profiles? How can I try to reproduce these profiles?
4. Do you FDO-optimize prebuilt binaries in your releases (e.g. available on GitHub Releases page)?

> Additionally, we would appreciate it if you could add our repository to the awesome repository related to FGO.

Of course! Just did it :)

**Comment by caifeizhi** [comment]
> 1\. Do you have the benchmark results of enabling FDO on OceanBase? I am interested in comparing Release vs Release + PGO OceanBase versions. So I will be able to put your benchmarks to my repo as an example.

about 15% improvement;

> 2\. I see you use Sampling (AutoFDO) approach. Did you try to use the Instrumented FDO instead? If yes, could you please describe why you chose AutoFDO instead? My guess is that Instrumented FDO has just too big performance overhead compared to AutoFDO but it's just my guess.
> 3\. How did you collect the committed to the repo FDO profiles? How can I try to reproduce these profiles?

@Naynahs can you please take a look

> 4\. Do you FDO-optimize prebuilt binaries in your releases (e.g. available on GitHub Releases page)?

yes, we use AUTOFDO=ON when rpm build, see it: https://github.com/oceanbase/oceanbase/blob/223cf90c93ddf0c163da2c47debe7ef4da2f46cd/build.sh#L207

**Comment by ruclz** [comment]
Yeah, I also want to know more about oceanbase, how to collect the committed to the repo FDO profiles? Thanks very much.
